src/combinatorics/simple_graph/subgraph.lean

```diff
@@ -0,0 +1,302 @@
+/-
+Copyright (c) 2021 Hunter Monroe. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Hunter Monroe, Kyle Miller, Alena Gusakov
+-/
+import combinatorics.simple_graph.basic
+import data.set.finite
+
+/-!
+# Subgraphs of a simple graph
+
+A subgraph of a simple graph consists of subsets of the graph's vertices and edges such that the
+endpoints of each edge are present in the vertex subset.  The edge subset is formalized as a
+sub-relation of the adjacency relation of the simple graph.
+
+## Main definitions
+
+* `subgraph G` is the type of subgraphs of a `G : simple_graph`
+
+* `subgraph.neighbor_set`, `subgraph.incidence_set`, and `subgraph.degree` are like their
+  `simple_graph` counterparts, but they refer to vertices from `G` to avoid subtype coercions.
+
+* `subgraph.coe` is the coercion from a `G' : subgraph G` to a `simple_graph G'.verts`.
+  (This cannot be a `has_coe` instance since the destination type depends on `G'`.)
+
+* A `bounded_lattice (subgraph G)` instance.
+
+* Graph homomorphisms from a subgraph to a graph (`subgraph.map_top`) and between subgraphs
+  (`subgraph.map`).
+
+## Implementation notes
+
+* Recall that subgraphs are not determined by their vertex sets, so `set_like` does not apply to
+  this kind of subobject.
+
+## Todo
+
+* Images of graph homomorphisms as subgraphs.
```

Comment:
It would be nice to have predicates saying that a subgraph is `induced` and `spanning`, probably for now they should be TODO rather than included in this PR though.

Reply:
For better or for worse, I went ahead and added some definitions for these.  I tried to make them so they give useful conclusions, but they'll still need lemmas for rewriting in the presence of these properties (though perhaps `is_induced` should have an `iff` conclusion so you can rewrite directly).